File: 0049-group-anagrams/0049-group-anagrams.kt
class Solution {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        val anagrams = HashMap<String, MutableList<String>>()

        for (word in strs) {
            val sorted = word.sorted()
            if (anagrams.containsKey(sorted)) anagrams[sorted]!!.add(word)
            else anagrams[sorted] = mutableListOf(word)
        }

        return anagrams.map {
            it.value.toList()
        }
    }

    private fun String.sorted(): String {
        val arr = this.toCharArray()
        arr.sort()
        return arr.joinToString("")
    }
}

File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> anagrams;

        for (auto &word : strs) {
            string sorted = word;
            sort(sorted.begin(), sorted.end());

            if (anagrams.find(sorted) != anagrams.end()) { 
                anagrams[sorted].push_back(word);
            }
            else {
                vector<string> newList = { word };
                anagrams[sorted] = newList;
            }
        }

        vector<vector<string>> result;
        for (auto &entry : anagrams) {
            result.push_back(entry.second);
        }
        return result;
    }
};

